options/ansi/include/stdlib.h

```diff
@@ -108,6 +108,12 @@ int wctomb(char *mb_chr, wchar_t wc);
 size_t mbstowcs(wchar_t *__restrict wc_string, const char *__restrict mb_string, size_t max_size);
 size_t wcstombs(char *mb_string, const wchar_t *__restrict wc_string, size_t max_size);
 
+#if __MLIBC_GLIBC_OPTION
+typedef int (*__compar_fn_t) (const void *, const void *);
```

Comment:
These typedefs are used in some `systemd` code, no use in libc as of this moment. Glibc defines them too.

Reply:
`__compar_fn_t` seems like a glibc internal typedef after looking at the source and man page. It's used for `tsearch` and friends to simplify the prototype, and shouldn't be used (it's only used by glibc internal files). `comparison_fn_t` seems like the public one (it's the one that's used in glibc tests). `__compar_d_fn_t` also seems like a convenience typedef that shouldn't be used by the user directly.

As a matter of fact, systemd code seems to only use `comparison_fn_t` out of the three (and mentions the last one in a comment when providing it's own typedef).

Reply:
Then I shall remove the other two and rebuild systemd to see if that assumption holds. Thanks for looking into it.

Reply:
Do note that `comparison_fn_t` should only be visible if the glibc option is enabled (and even in glibc headers it's guarded by `_USE_GNU`).

Reply:
We already guard it behind the glibc option, the `_USE_GNU` defines (and its friends) are something we do not (yet) use in mlibc, and as such I'm not guarding anything behind that.

Reply:
systemd disagrees on the fact that `__compar_fn_t` and `__compar_d_fn_t` aren't used, see build log below.

```
In file included from ../../../src/ports/systemd/src/udev/cdrom_id/cdrom_id.c:17:
../../../src/ports/systemd/src/basic/sort-util.h:9:18: error: unknown type name ‘__compar_d_fn_t’
    9 |                  __compar_d_fn_t compar, void *arg);
      |                  ^~~~~~~~~~~~~~~
../../../src/ports/systemd/src/basic/sort-util.h:23:61: error: unknown type name ‘__compar_fn_t’
   23 |                                  size_t nmemb, size_t size, __compar_fn_t compar) {
      |                                                             ^~~~~~~~~~~~~
../../../src/ports/systemd/src/basic/sort-util.h:42:71: error: unknown type name ‘__compar_fn_t’
   42 | static inline void _qsort_safe(void *base, size_t nmemb, size_t size, __compar_fn_t compar) {
      |                                                                       ^~~~~~~~~~~~~
../../../src/ports/systemd/src/basic/sort-util.h:58:72: error: unknown type name ‘__compar_d_fn_t’
   58 | static inline void qsort_r_safe(void *base, size_t nmemb, size_t size, __compar_d_fn_t compar, void *userdata) {
      |                                                                        
```
As such, I propose letting them stay.

Reply:
What version of systemd are you trying? The change to avoid using glibc-internal types was made on 11 Oct 2021: https://github.com/systemd/systemd/commit/d8f16737005eb80b34df98d96f61f0b8542b098f (and the parent of this commit).

Reply:
Hmm. It should be 252 but I guess it's 249 still then. I'll recheck later and make sure I'm on 252.

Reply:
Rechecked, no longer needed. Dropped the offending lines from this PR. Good catch!